Don't assume TypeBuilder types can be emitted as tokens.

JonHanna opened the pull request:
Fixes #15852

stephentoub commented on src/System.Linq.Expressions/tests/Constant/ConstantTests.cs:
```diff
@@ -276,7 +297,17 @@ public static void CheckUShortConstantTest(bool useInterpreter)
+                typeof(int),
+                typeof(Func<string>),
+                typeof(List<>).GetGenericArguments()[0],
+                GetTypeBuilder(),
```
So this fails before the fix and passes after, I assume?

JonHanna replied:
Yes. `ArgumentException` with a message of "Type must be a runtime Type object."

stephentoub commented on src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/ILGen.cs:
```diff
@@ -574,10 +574,7 @@ internal static void EmitConstant(this ILGenerator il, object value, Type type)
-        {
-            return t is TypeBuilder || t.IsGenericParameter || t.GetTypeInfo().IsVisible;
-        }
+        private static bool ShouldLdtoken(Type t) => t.IsGenericParameter || t.GetTypeInfo().IsVisible;
```
Are there any situations in which the t is TypeBuilder clause is valid?  I'm wondering if the right answer is to remove it or if the right answer is to further qualify it?

JonHanna replied:
I've been wondering the same thing. I think the limitation might be with dynamic methods and CompileToMethod would be okay, which would be worth checking if that gets brought to core, especially since CompileToMethod is less flexible about the fallback dynamic methods use when ShouldLdtoken fails.

JonHanna replied:
Yes, that's it. The following works in LINQPad and returns the created type, rather than the TypeBuilder used to create it:

```C#
AssemblyBuilder assembly = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("Name"), AssemblyBuilderAccess.Run);
ModuleBuilder module = assembly.DefineDynamicModule("Name");
var typeBuilder = module.DefineType("Type");
var meth = typeBuilder.DefineMethod("r", MethodAttributes.Public | MethodAttributes.Static, typeof(TypeBuilder), Type.EmptyTypes);
var e = Expression.Lambda<Func<Type>>(Expression.Constant(typeBuilder, typeof(Type)));
e.CompileToMethod(meth);
var t = typeBuilder.CreateTypeInfo().AsType();
object result = t.GetMethod("r").Invoke(null, new object[0]);
LINQPad.Extensions.Dump(result);
```

Need to allow that to continue.

JonHanna commented:
There's now a check on whether the method is a DynamicMethod, which is always true when `FEATURE_COMPILE_TO_METHODBUILDER` isn't set. That should stop this from introducing the opposite issue if `CompileToMethod` is made available.

JonHanna commented:
Test Innerloop OSX Debug Build and Test Please
(Error restoring build tools. Will open issue if it repeats.)

JonHanna commented:
@VSadov @bartdesmet you discussed bringing `CompileToMethod` back. What are the odds of this happening? If it definitely isn't happen then all that's needed here is to remove the `t is TypeBuilder ||` and this PR can be a lot simpler.

JonHanna commented:
I've simplified back to just leaving out accepting `TypeBuilder` as an IL constant and added a comment for the possibility of CompileToMethod coming back. Otherwise it's a lot of passing state around for something that might not be enabled.

JonHanna commented:
Test Innerloop Windows_NT Release Build and Test

VSadov commented:
@JonHanna agreed. Just removing it for now seems simpler.
